src/BasicVault.sol

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "../lib/yield-utils-v2/contracts/token/IERC20.sol";
+import "../lib/yield-utils-v2/contracts/token/IERC2612.sol";
+
+
+/// @title Basic Vault
+/// @author hashedMae
+/// @notice A simple vault that tracks the amount of tokens a user has deposited and allows them to withdraw at any time
+/// @dev 
+
+contract BasicVault {
+
+IERC20 iToken;
+
+mapping(address => uint256) public userBalance;
+
+/// @notice Emitted when a user deposits token to the contract
+event Deposit(address indexed user, uint256 amount, uint256 balance);
+/// @notice Emitted when a user withdraws tokens from the contract
+event Withdrawal(address indexed user, uint256 amount, uint256 balance);
+
+/// @param iToken_ The address of the token that the vault will hold
+constructor(IERC20 iToken_){
+    iToken = iToken_;
+}
+
+/// @notice Allows a user to deposit tokens to the vault
+/// @dev Tracks user vault balance via mapping of address => uint
+/// @param amount Number of Tokens to deposit
+function depositToken(uint256 amount) public {
+    require(iToken.transferFrom(msg.sender, address(this), amount), "failed on tranfer");
+    userBalance[msg.sender] += amount;
+    emit Deposit(msg.sender, amount, userBalance[msg.sender]);
+}
+
+/// @notice Allows a user to withdraw tokens they've previously deposited
+/// @dev ensures that a user can't withdraw more tokens than they've deposited, subtracts from user's balance after succesful withdrawal
+/// @param amount Number of tokens to withdraw
+function withdrawToken(uint256 amount) public {
+    require(userBalance[msg.sender] >= amount, "requested amount exceeds user balance");
+    require(iToken.transferFrom(address(this), msg.sender, amount), "failed on withdrawal");
+    userBalance[msg.sender] -= amount;
```

Comment:
If you are certain of the implementation of `iToken` this line and the previous one would be fine. However, if the user would be somehow allowed to register new `iToken` (a kind a multi-token user-defined vault, not crazy) this ordering would open your contract to a reentrancy attack.

`iToken.transferFrom` could have been maliciously coded to call `withdrawToken` again, and the contract would transfer iToken out in a loop without modifying the user balance. The attacker would only have to be careful of stopping the loop before running out of gas, and it would drain your contract of iToken.

Please Google about reentrancy, the TheDAO hack, and the checks-effects-interactions pattern.